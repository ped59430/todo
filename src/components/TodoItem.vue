<template>
  <div>
    <q-item v-if="editMode">
      <div class="full-width row">
        <q-input
          ref="textinput"
          @keydown.esc="setEditFalse"
          @blur="setEditFalse"
          @keydown.enter="onEnter"
          class="col-xs-8 col-md-10"
          filled
          v-model="inputText"
          autofocus
        />
        <q-btn class="col-xs-4 col-md-2 flex-center" @click="onEnter"
          >Submit</q-btn
        >
      </div>
    </q-item>
    <q-item clickable class="items-center" v-else>
      <div class="full-width row flex items-center">
        <div class="col-xs-2 col-md-1">
          <q-checkbox v-model="checkbox" />
        </div>
        <span
          :style="done && 'text-decoration: line-through'"
          @click="setEditTrue"
          class="col-xs-8 col-md-10"
          >{{ description }}</span
        >
        <q-btn class="col-xs-2 col-md-1 bg-red flex-center" @click="deleteTodo">
          <q-icon color="white" name="delete"></q-icon>
        </q-btn>
      </div>
    </q-item>
  </div>
</template>

<script>
export default {
  name: "TodoItem",
  props: {
    uid: { type: Number, required: true },
    description: { type: String, required: true },
    done: { type: Boolean, required: true }
  },
  data() {
    return {
      editMode: false,
      inputText: ""
    };
  },
  methods: {
    editTodo(todo) {
      return this.$store.dispatch("todo/serverEditTodo", todo);
    },
    deleteTodo() {
      return this.$store.dispatch("todo/serverDeleteTodo", this.uid);
    },
    setEditTrue() {
      this.editMode = true;
    },
    setEditFalse() {
      this.editMode = false;
    },
    async onEnter() {
      await this.editTodo({
        uid: this.uid,
        description: this.inputText,
        done: this.done
      });
      this.editMode = false;
    }
  },
  computed: {
    checkbox: {
      get: function() {
        return this.done;
      },
      set: function() {
        this.editTodo({
          uid: this.uid,
          description: this.description,
          done: !this.done
        });
      }
    }
  },
  mounted() {
    this.inputText = this.description;
  }
};
</script>

<style></style>
