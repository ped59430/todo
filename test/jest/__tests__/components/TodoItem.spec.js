/* eslint-disable */
/**
 * @jest-environment jsdom
 */

import { mount, createLocalVue, shallowMount } from "@vue/test-utils";
import * as All from "quasar";
import Vuex from "vuex";
import TodoItem from "~/src/components/TodoItem.vue";
const { Quasar } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key];
  if (val && val.component && val.component.name != null) {
    object[key] = val;
  }
  return object;
}, {});

describe("TodoItem", () => {
  const localVue = createLocalVue();
  localVue.use(Quasar, { components });
  localVue.use(Vuex);

  let store;
  let mutations;
  let wrapper;
  let actions;
  const todo = {
    uid: 0,
    description: "zero",
    done: false
  };

  beforeAll(() => {
    mutations = {
      MUTATION: jest.fn()
    };
    actions = {
      serverEditTodo: jest.fn(),
      serverDeleteTodo: jest.fn()
    };
    store = new Vuex.Store({
      modules: {
        todo: {
          namespaced: true,
          state: {},
          mutations,
          actions
        }
      }
    });
    wrapper = mount(TodoItem, {
      store,
      localVue,
      propsData: todo
    });
  });

  it("snapshot testing", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("setEdit change editMode properly", () => {
    const { vm } = wrapper;
    vm.setEditTrue();
    expect(vm.editMode).toEqual(true);
    vm.setEditFalse();
    expect(vm.editMode).toEqual(false);
  });
});
