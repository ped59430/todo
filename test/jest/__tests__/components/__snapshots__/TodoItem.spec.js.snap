// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem snapshot testing 1`] = `
<div>
  <div tabindex="0" class="q-item q-item-type row no-wrap items-center q-item--clickable q-link cursor-pointer q-focusable q-hoverable">
    <div tabindex="-1" class="q-focus-helper"></div>
    <div class="full-width row flex items-center">
      <div class="col-xs-2 col-md-1">
        <div tabindex="0" class="q-checkbox cursor-pointer no-outline row inline no-wrap items-center">
          <div class="q-checkbox__inner relative-position"><input type="checkbox" class="q-checkbox__native q-ma-none q-pa-none invisible">
            <div class="q-checkbox__bg absolute"><svg viewBox="0 0 24 24" class="q-checkbox__check fit absolute-full">
                <path fill="none" d="M1.73,12.91 8.1,19.28 22.79,4.59"></path>
              </svg>
              <div class="q-checkbox__check-indet absolute"></div>
            </div>
          </div>
        </div>
      </div> <span class="col-xs-8 col-md-10">zero</span> <button tabindex="0" type="button" class="q-btn inline q-btn-item non-selectable col-xs-2 col-md-1 bg-red flex-center q-btn--standard q-btn--rectangle q-focusable q-hoverable">
        <div tabindex="-1" class="q-focus-helper"></div>
        <div class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i aria-hidden="true" class="material-icons q-icon text-white">delete</i></div>
      </button>
    </div>
  </div>
</div>
`;
